src/OrchardCore/OrchardCore.Abstractions/Shell/Scope/ShellScope.cs

```diff
@@ -52,6 +52,15 @@ public ShellScope(ShellContext shellContext)
 
             _serviceScope = shellContext.ServiceProvider.CreateAsyncScope();
             ServiceProvider = _serviceScope.ServiceProvider;
+
+            _beforeDispose = [
+                async scope =>
+                {
+                    var shellContextReleaseService = scope.ServiceProvider.GetRequiredService<IDeferredShellContextReleaseService>();
+
+                    await shellContextReleaseService.ProcessAsync();
+                }
+            ];
```

Comment:
This (and the `_beforeDispose` change` isn't needed. You can handle all of this from the service:

```c#
public class DeferredShellContextReleaseService : IDeferredShellContextReleaseService
{
    private bool _release;
    private bool _deferredTaskAdded;

    public void SuspendReleaseRequest()
    {
        _release = false;
    }

    public void RequestRelease()
    {
        _release = true;

        if (!_deferredTaskAdded) return;

        _deferredTaskAdded = true;

        ShellScope.AddDeferredTask(async scope =>
        {
            if (!_release) return;

            await scope.ServiceProvider
                .GetRequiredService<IShellHost>()
                .ReleaseShellContextAsync(scope.ServiceProvider.GetRequiredService<ShellSettings>());
        });
    }
}
```

Reply:
I'll fix the logical error and try it. Calling the _release property from the deferred functions seems weird to me.

Reply:
@Piedone this suggestion worked. It's also better than what I had. It's still weird to call `_release` from a callback like that but it works and simplify things